activerecord/lib/active_record/relation/batches.rb

```diff
@@ -373,32 +373,44 @@ def batch_on_unloaded_relation(relation:, start:, finish:, load:, order:, use_ra
           if load
             records = batch_relation.records
             ids = records.map(&:id)
+            ids_size = ids.size
+            ids_last = ids.last
             yielded_relation = where(primary_key => ids)
             yielded_relation.load_records(records)
           elsif (empty_scope && use_ranges != false) || use_ranges
-            ids = batch_relation.ids
-            finish = ids.last
-            if finish
-              yielded_relation = apply_finish_limit(batch_relation, finish, batch_orders)
+            # Efficiently peak at the last id for the next batch using offset and limit.
+            ids_size = batch_limit
+            ids_last = batch_relation.offset(batch_limit - 1).limit(1).ids.first
+            # If the last id is not found using offset, there is at most one more batch of size < batch_limit.
+            # We retry getting the list of ids so that we have the exact size and last id.
+            unless ids_last
+              ids = batch_relation.ids
+              ids_size = ids.size
+              ids_last = ids.last
+            end
```

Comment:
Up to you, but this conditional can be avoided by making the `else` branch for `if ids_last` with something like
```ruby
else
  ids_size = nil
end
```

and then instead of checking `break if ids_size == 0` do `break unless ids_size`.

Reply:
The code as currently implemented goes through each of these steps in order:

1. Find batch size and last id using OFFSET and LIMIT (fast path)
2. When 1. failed (no result), retry by querying the whole range of ids (slow path)
3. If either 1. of 2. was able to return at least one id, build the yielded relation and actually yield it

If I understand your comment correctly, we would try to remove step 2. and add an `else` branch to step 3.

I don't see how this could work, but maybe I misunderstood your proposal?

Reply:
Note: this snippet is a bit outdated since this branch has been updated to comply with https://github.com/rails/rails/pull/52384.

But the logic stays the same, so I still don't see how I could change the implementation with breaking the logic.